pkg/analyze/archivereader.go

```diff
@@ -0,0 +1,102 @@
+package analyze
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"reflect"
+
+	scyllav1 "github.com/scylladb/scylla-operator/pkg/api/scylla/v1"
+	scyllav1listers "github.com/scylladb/scylla-operator/pkg/client/scylla/listers/scylla/v1"
+	"github.com/scylladb/scylla-operator/pkg/scheme"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+func getIndexerForType(indexers map[reflect.Type]cache.Indexer, objType reflect.Type) cache.Indexer {
+	if indexer, exists := indexers[objType]; exists {
+		return indexer
+	}
+	// Create new indexer for this type if it doesn't exist
+	indexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{})
+	indexers[objType] = indexer
+	return indexer
+}
+
+func IndexersFromFS(fsys fs.FS, decoder runtime.Decoder) (map[reflect.Type]cache.Indexer, error) {
+	indexers := make(map[reflect.Type]cache.Indexer)
+
+	err := fs.WalkDir(fsys, ".", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() || filepath.Ext(path) != ".yaml" {
+			return nil
+		}
+		content, err := fs.ReadFile(fsys, path)
+		if err != nil {
+			return fmt.Errorf("can't read file %q: %w", path, err)
+		}
+		obj, _, err := decoder.Decode(content, nil, nil)
+
+		if err != nil {
+			if !runtime.IsNotRegisteredError(err) {
+				return fmt.Errorf("can't deserialize file %q: %w", path, err)
+			}
+			return nil
+		}
+		objType := reflect.TypeOf(obj)
+		indexer := getIndexerForType(indexers, objType)
+		err = indexer.Add(obj)
+		if err != nil {
+			return fmt.Errorf("can't add object to indexer: %w", err)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("can't walk the file tree: %w", err)
+	}
+	return indexers, nil
+}
+
+func NewDataSourceFromFS(ctx context.Context, archivePath string, decoder runtime.Decoder) (*DataSource, error) {
+	if decoder == nil {
+		decoder = serializer.NewCodecFactory(scheme.Scheme).UniversalDecoder()
+	}
+
+	indexers, err := IndexersFromFS(os.DirFS(archivePath), decoder)
+	if err != nil {
+		return nil, fmt.Errorf("can't build listers from fs: %w", err)
+	}
+
+	types := []interface{}{
+		&corev1.Pod{},
+		&corev1.Service{},
+		&corev1.Secret{},
+		&corev1.ConfigMap{},
+		&corev1.ServiceAccount{},
+		&scyllav1.ScyllaCluster{},
+	}
+
+	for _, obj := range types {
+		t := reflect.TypeOf(obj)
+		if indexers[t] == nil {
+			indexers[t] = cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{})
+		}
+	}
```

Comment:
you already have the `getIndexerForType` helper, so you could also use that when creating listers